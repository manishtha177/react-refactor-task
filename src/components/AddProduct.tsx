import React from 'react';
import Modal from "react-modal";
import styles from "../shopApp.module.css";
import { FaTimes } from "react-icons/fa";
import { Form } from "../components/form";

interface IAddProductProps {
	isOpen: boolean;
	toggleAddProductModal: () => void;
	onSubmit: (payload: { title: string; description: string; price: string }) => void;
}

const AddProduct: React.FC<IAddProductProps> = ({ isOpen, toggleAddProductModal, onSubmit }) => {
	return (
		<Modal
			isOpen={isOpen}
			className={styles.reactModalContent}
			overlayClassName={styles.reactModalOverlay}
			onRequestClose={toggleAddProductModal}
		>
			<div className={styles.modalContentHelper}>
				<div
					className={styles.modalClose}
					onClick={toggleAddProductModal}>
					<FaTimes />
				</div>

				<Form on-submit={onSubmit} />
			</div>
		</Modal>
	)
}

export default AddProduct