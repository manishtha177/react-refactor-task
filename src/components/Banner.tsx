import React from "react";
import img1 from "../images/img1.png";
import img2 from "../images/img2.png";
import styles from "../shopApp.module.css";

const Banner = () => {
  return (
    <>
      <span className={`container ${styles.main} ${styles.images}`}>
        <img data-testid={"img-1"} src={img1} alt="docter" />
        <img data-testid={"img-2"} src={img2} alt="engineer" />
      </span>
    </>
  );
};

export default Banner;
