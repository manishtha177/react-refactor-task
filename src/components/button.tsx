import * as React from "react";
import { IButtonProps } from "../interfaces/button";
import styles from "./button.module.css";

export const Button: React.FC<IButtonProps> = ({ children, onClick }) => (
  <button className={styles.button} onClick={onClick}>
    {children}
  </button>
);
