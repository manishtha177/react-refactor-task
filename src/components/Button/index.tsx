import * as React from "react";
import { IButtonProps } from "../../interfaces/button";
import styles from "./index.module.css";

export const Button: React.FC<IButtonProps> = ({ children, onClick }: IButtonProps) => (
  <button className={styles.button} onClick={onClick} data-testid={"send-product-button"}>
    {children}
  </button>
);
