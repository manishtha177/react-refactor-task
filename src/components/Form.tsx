import React, { useState, ChangeEvent, SyntheticEvent } from "react";
import { IFormData, IFormError, IFormProps } from "../interfaces/form";
import { validateForm } from "../utils/validate";
import { Button } from "./Button";
import styles from "./Form.module.css";

export const Form: React.FC<IFormProps> = (props) => {
  const [formData, setFormData] = useState<IFormData>({
    title: "",
    price: 0,
    description: "",
  });
  const [error, setError] = useState<IFormError>({
    title: false,
    price: false,
    description: false,
  });

  const handleChange = (
    e: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
  ) => {
    e.persist();
    setError((prevError) => ({
      ...prevError,
      [e.target.name]: false,
    }));
    setFormData((prevData) => ({
      ...prevData,
      [e.target.name]: e.target.value,
    }));
  };

  const handleSubmit = (e: SyntheticEvent) => {
    e.preventDefault();
    const { formIsValid, errors } = validateForm(formData, error);
    if (formIsValid) {
      props.onSubmit({
        title: formData.title,
        description: formData.description,
        price: formData.price,
      });

      setFormData({ title: "", price: 0, description: "" });
    } else {
      console.log("errors : ", errors);
      setError(errors);
    }
  };

  return (
    <form className={styles.form} onSubmit={(event) => handleSubmit(event)}>
      <span className={styles.label}>Product title: *</span>

      <input
        placeholder="Title..."
        name="title"
        value={formData.title}
        className={styles.input}
        onChange={handleChange}
      />
      {error.title && (
        <span className={styles.error}>Product Title is required</span>
      )}

      <span className={styles.label}>Product Price: *</span>

      <input
        placeholder="Price..."
        type="number"
        name="price"
        value={formData.price}
        className={styles.input}
        onChange={handleChange}
      />
      {error.price && (
        <span className={styles.error}>Product Price is required</span>
      )}

      <span className={styles.label}>Product Description: *</span>

      <textarea
        placeholder="Start typing product description here..."
        name="description"
        value={formData.description}
        className={styles.textarea}
        onChange={handleChange}
      />
      {error.description && (
        <span className={styles.error}>Product Description is required</span>
      )}

      <div className={styles.buttonWrapper}>
        <Button>Add a product</Button>
      </div>
    </form>
  );
};
