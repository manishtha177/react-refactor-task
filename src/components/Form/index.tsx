import React, { useState, ChangeEvent, SyntheticEvent } from "react";
import {
  IFormData,
  IFormError,
  IFormProps,
  IField,
} from "../../interfaces/form";
import { constants } from "../../utils/constants";
import { validateForm } from "../../utils/validate";
import { Button } from "../Button";
import styles from "./index.module.css";
import Input from "../Input";

export const Form: React.FC<IFormProps> = (props: IFormProps) => {
  const formDummyData = { title: "", price: 0, description: "" };
  const [formData, setFormData] = useState<IFormData>(formDummyData);

  const [error, setError] = useState<IFormError>({
    title: false,
    price: false,
    description: false,
  });

  const handleChange = (
    e: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
  ) => {
    let name = e?.target?.name;
    let value = e?.target?.value;
    e.persist();
    setError((prevError) => ({
      ...prevError,
      [name]: false,
    }));
    if ("price" === name) {
      value = value[0] === "0" ? value[1] : value;
    }
    setFormData((prevData) => ({
      ...prevData,
      [name]: value,
    }));
  };

  const handleSubmit = (e: SyntheticEvent) => {
    e.preventDefault();
    const { formIsValid, errors } = validateForm(formData, error);
    if (formIsValid) {
      props.onSubmit({
        title: formData?.title,
        description: formData?.description,
        price: formData?.price,
      });

      setFormData(formDummyData);
    } else {
      console.log("errors : ", errors);
      setError(errors);
    }
  };

  const fields: IField[] = [
    {
      name: "title",
      label: constants.PRODUCT_TITLE,
      type: "text",
      placeHolder: "Title...",
      value: formData?.title,
      onChange: handleChange,
    },
    {
      name: "price",
      label: constants.PRODUCT_PRICE,
      type: "number",
      placeHolder: "Price...",
      value: formData?.price,
      onChange: handleChange,
    },
  ];

  return (
    <form
      className={styles.form}
      onSubmit={(event) => handleSubmit(event)}
      data-testid={"form"}
    >
      {fields.map((field) => (
        <Input key={field.name} field={field} error={error} />
      ))}

      <span className={styles.label}>{constants.PRODUCT_DESCRIPTION}: *</span>

      <textarea
        placeholder="Start typing product description here..."
        name="description"
        value={formData?.description}
        className={styles?.textarea}
        onChange={handleChange}
      />
      <p className={styles.error}>
        {error?.description && `${constants.PRODUCT_DESCRIPTION} is required`}
      </p>

      <div className={styles.buttonWrapper}>
        <Button>Add a product</Button>
      </div>
    </form>
  );
};
