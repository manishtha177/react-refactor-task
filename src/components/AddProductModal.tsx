import React from "react";
import Modal from "react-modal";
import styles from "../shopApp.module.css";
import { FaTimes } from "react-icons/fa";
import { Form } from "./Form";
import { IAddProductProps } from "../interfaces/add-product";

const AddProductModal: React.FC<IAddProductProps> = ({
  isOpen,
  toggleAddProductModal,
  onSubmit,
}: IAddProductProps) => {
  return (
    <Modal
      isOpen={isOpen}
      className={styles.reactModalContent}
      overlayClassName={styles.reactModalOverlay}
      onRequestClose={toggleAddProductModal}
    >
      <div className={styles.modalContentHelper}>
        <div className={styles.modalClose} onClick={toggleAddProductModal}>
          <FaTimes />
        </div>

        <Form onSubmit={onSubmit} />
      </div>
    </Modal>
  );
};

export default AddProductModal;
