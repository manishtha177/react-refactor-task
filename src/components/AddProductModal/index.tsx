import React from "react";
import Modal from "react-modal";
import styles from "../../Pages/ShopApp/index.module.css";
import { FaTimes } from "react-icons/fa";
import { Form } from "../Form/index";
import { IAddProductProps } from "../../interfaces/add-product";
import { constants } from "../../utils/constants";
import { addProduct } from "../../services/addProduct";

const AddProductModal: React.FC<IAddProductProps> = ({
  isOpen,
  toggleAddProductModal,
  setData,
  shopData,
}: IAddProductProps) => {
  const onSubmit = async (payload: {
    title: string;
    description: string;
    price: number;
  }) => {
    setData({
      isOpen: false,
      isShowingMessage: true,
      message: constants.PRODUCT_ADDING_MESSAGE,
    });

    try {
      const response = await addProduct(payload);
      if (response?.id) {
        const tempProducts = shopData?.products;
        tempProducts.unshift({
          id: response?.id,
          title: payload?.title,
          description: payload?.description,
          price: payload?.price,
          isFavorite: false,
          rating: { rate: 0, count: 0 },
        });
        setData({
          products: tempProducts,
          isShowingMessage: true,
          message: constants.PRODUCT_ADDDED_SUCCESSFULLY_MESSAGE,
        });
      } else {
        setData({
          isOpen: false,
          isShowingMessage: true,
          message: constants.FAILED_TO_ADD_PRODUCT_MESSAGE,
        });
      }
    } catch (error) {
      console.log("error : ", error);
      setData({
        isOpen: false,
        isShowingMessage: true,
        message: constants.SOMETHING_WENT_WRONT_MESSAGE,
      });
    } finally {
      setTimeout(() => {
        setData({ isShowingMessage: false, message: "" });
      }, 2000);
    }
  };

  return (
    <Modal
      isOpen={isOpen}
      className={styles.reactModalContent}
      overlayClassName={styles.reactModalOverlay}
      onRequestClose={toggleAddProductModal}
    >
      <div className={styles.modalContentHelper}>
        <div className={styles.modalClose} onClick={toggleAddProductModal}>
          <FaTimes />
        </div>

        <Form onSubmit={onSubmit} />
      </div>
    </Modal>
  );
};

export default AddProductModal;
