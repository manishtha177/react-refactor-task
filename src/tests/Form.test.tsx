import * as React from "react";
import { render, screen } from "@testing-library/react";
import { Form } from "../components/Form";

test("render form & take snapshot for correct render", () => {
  render(<Form onSubmit={() => console.log("submit click")} />);
  expect(screen.getByTestId("form")).toMatchSnapshot();
  expect(screen.getByTestId("form")).toBeTruthy();
});

test("render Product Title input field", () => {
  render(<Form onSubmit={() => console.log("submit click")} />);
  const titleElement = screen.getByPlaceholderText("Title...");
  expect(titleElement).toHaveValue("");
});

test("render Product Price input field", () => {
  render(<Form onSubmit={() => console.log("submit click")} />);
  const priceElement = screen.getByPlaceholderText("Price...");
  expect(priceElement).toHaveValue(0);
});

test("render Product Description textarea field", () => {
  render(<Form onSubmit={() => console.log("submit click")} />);
  const descriptionElement = screen.getByPlaceholderText(
    "Start typing product description here..."
  );
  expect(descriptionElement).toHaveValue("");
});
