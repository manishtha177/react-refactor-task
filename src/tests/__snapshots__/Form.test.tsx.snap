// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correctly 1`] = `
<form
  class="form"
  data-testid="form"
>
  <div>
    <span
      class="label"
    >
      Product Title
       *
    </span>
    <input
      class="input"
      name="title"
      placeholder="Title..."
      type="text"
      value=""
    />
    <p
      class="error"
    />
  </div>
  <div>
    <span
      class="label"
    >
      Product Price
       *
    </span>
    <input
      class="input"
      name="price"
      placeholder="Price..."
      type="number"
      value="0"
    />
    <p
      class="error"
    />
  </div>
  <span
    class="label"
  >
    Product Description
    : *
  </span>
  <textarea
    class="textarea"
    name="description"
    placeholder="Start typing product description here..."
  />
  <p
    class="error"
  />
  <div
    class="buttonWrapper"
  >
    <button
      class="button"
      data-testid="send-product-button"
    >
      Add a product
    </button>
  </div>
</form>
`;

exports[`render form & take snapshot for correct render 1`] = `
<form
  class="form"
  data-testid="form"
>
  <div>
    <span
      class="label"
    >
      Product Title
       *
    </span>
    <input
      class="input"
      name="title"
      placeholder="Title..."
      type="text"
      value=""
    />
    <p
      class="error"
    />
  </div>
  <div>
    <span
      class="label"
    >
      Product Price
       *
    </span>
    <input
      class="input"
      name="price"
      placeholder="Price..."
      type="number"
      value="0"
    />
    <p
      class="error"
    />
  </div>
  <span
    class="label"
  >
    Product Description
    : *
  </span>
  <textarea
    class="textarea"
    name="description"
    placeholder="Start typing product description here..."
  />
  <p
    class="error"
  />
  <div
    class="buttonWrapper"
  >
    <button
      class="button"
      data-testid="send-product-button"
    >
      Add a product
    </button>
  </div>
</form>
`;

exports[`render form 1`] = `
<form
  class="form"
  data-testid="form"
>
  <div>
    <span
      class="label"
    >
      Product Title
       *
    </span>
    <input
      class="input"
      name="title"
      placeholder="Title..."
      type="text"
      value=""
    />
    <p
      class="error"
    />
  </div>
  <div>
    <span
      class="label"
    >
      Product Price
       *
    </span>
    <input
      class="input"
      name="price"
      placeholder="Price..."
      type="number"
      value="0"
    />
    <p
      class="error"
    />
  </div>
  <span
    class="label"
  >
    Product Description
    : *
  </span>
  <textarea
    class="textarea"
    name="description"
    placeholder="Start typing product description here..."
  />
  <p
    class="error"
  />
  <div
    class="buttonWrapper"
  >
    <button
      class="button"
      data-testid="send-product-button"
    >
      Add a product
    </button>
  </div>
</form>
`;
